webpack/set-webpack-public-path-plugin/src/SetPublicPathPlugin.ts

```diff
@@ -127,48 +147,79 @@ export class SetPublicPathPlugin implements Webpack.Plugin {
     this.options = options;
+
+    if (options.scriptName) {
+      if (options.scriptName.useAssetName && options.scriptName.name) {
+        throw new Error('scriptName.userAssetName and scriptName.name must not be used together');
+      } else if (options.scriptName.isTokenized && !options.scriptName.name) {
+        throw new Error('scriptName.isTokenized is only valid if scriptName.name is set');
+      }
+    }
   }
 
   public apply(compiler: Webpack.Compiler): void {
     const isWebpack4: boolean = !!compiler.hooks;
 
     if (isWebpack4) {
-      compiler.hooks.compilation.tap('set-webpack-public-path', (compilation: Webpack.compilation.Compilation) => {
+      compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation: Webpack.compilation.Compilation) => {
         const v4MainTemplate: IV4MainTemplate = compilation.mainTemplate as IV4MainTemplate;
-        v4MainTemplate.hooks.startup.tap(
-          'set-webpack-public-path',
-          (source: string, chunk: IV4Chunk, hash: string) => {
-            let assetOrChunkFound: boolean = !!this.options.skipDetection;
-
-            if (!assetOrChunkFound) {
-              for (const chunkGroup of chunk.groupsIterable) {
-                const children: Webpack.compilation.Chunk[] = chunkGroup.getChildren();
-                assetOrChunkFound = assetOrChunkFound || (children.length > 0);
-              }
+        v4MainTemplate.hooks.startup.tap(PLUGIN_NAME, (source: string, chunk: IV4Chunk, hash: string) => {
+          let assetOrChunkFound: boolean = !!this.options.skipDetection;
+
+          if (!assetOrChunkFound) {
+            for (const chunkGroup of chunk.groupsIterable) {
+              const children: Webpack.compilation.Chunk[] = chunkGroup.getChildren();
+              assetOrChunkFound = assetOrChunkFound || (children.length > 0);
             }
+          }
 
-            if (!assetOrChunkFound) {
-              for (const innerModule of chunk.modulesIterable) {
-                if (innerModule.buildInfo.assets && Object.keys(innerModule.buildInfo.assets).length > 0) {
-                  assetOrChunkFound = true;
-                }
+          if (!assetOrChunkFound) {
+            for (const innerModule of chunk.modulesIterable) {
+              if (innerModule.buildInfo.assets && Object.keys(innerModule.buildInfo.assets).length > 0) {
+                assetOrChunkFound = true;
               }
             }
+          }
+
+          if (assetOrChunkFound) {
+            return this._getStartupCode({
+              source,
+              chunk,
+              hash,
+              requireFn: v4MainTemplate.requireFn
+            });
+          } else {
+            return source;
+          }
+        });
+      });
+
+      compiler.hooks.emit.tap(PLUGIN_NAME, (compilation: Webpack.compilation.Compilation) => {
+        for (const chunkGroup of compilation.chunkGroups) {
+          for (const chunk of chunkGroup.chunks) {
```

Comment:
can this be optimized?

Reply:
Same story - we still need to loop through each chunk-produced asset, so regardless of how we loop through them, it'll be the same number of elements.